storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1745 @@ buf_block_t *fsp_get_latched_xdes_page(
+                            mtr);
+
+    associatedPages related_pages(block, mtr);
+    err= allocator.prepare(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) return err;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) return err;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr->discard_modifications();
+      return err;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    /* Copy FIL_PAGE_PREV, FIL_PAGE_NEXT */
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_PREV,
+                block->page.frame + FIL_PAGE_PREV, 8);
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_TYPE,
+                block->page.frame + FIL_PAGE_TYPE,
+                srv_page_size - FIL_PAGE_TYPE - 8);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) return err;
+    }
+    goto fetch_next_page;
+  }
+
+  ut_a(!fsp_tablespace_validate(space, mtr));
+  ut_a(!fseg_validate(space, m_index, mtr));
+  if (level > 1)
+  {
+    mtr->commit();
+    mtr->start();
+    mtr->x_lock_space(space);
+  }
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag) noexcept
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  m_index->lock.x_lock(SRW_LOCK_CALL);
+  fil_space_t *const space= fil_system.sys_space;
+  mtr.start();
+  mtr.x_lock_space(space);
+  m_root= btr_root_block_get(m_index, RW_S_LATCH, &mtr, &err);
+  if (!m_root)
+  {
+    mtr.commit();
+    m_index->lock.x_unlock();
+    return err;
+  }
+
+  m_root->page.fix();
+  mtr.release_last_page();
+  uint16_t level= btr_page_get_level(m_root->page.frame);
+  while (1)
+  {
+    err= defragment_level(level, &mtr, space_defrag);
+    DBUG_EXECUTE_IF("fail_after_level_defragment",
+                    if (level == 1)
+                      err= DB_CORRUPTION;);
+    if (err || !level)
+      break;
+    level--;
+  }
+  mtr.commit();
```

Comment:
Before the `mtr.commit()` invocation, please add a debug assertion to ensure that any changes will be discarded in case any error occurred:
```c++
  ut_ad(err == DB_SUCCESS || !mtr.has_modifications());
```